File: frontend/lib/types/contracts.ts
export type ContextSourceType = 'FILE' | 'DIRECTORY' | 'URL' | 'API' | 'DATABASE' | 'MEMORY' | 'OBSIDIAN';

export interface ContextSourceContract {
  type: ContextSourceType;
  path: string;
  extra?: Record<string, any>;
}

export type ObsidianEventType = 'open' | 'click' | 'selection' | 'navigate';

export interface ObsidianContextEventContract {
  event_id?: string;
  event_type: ObsidianEventType;
  note_path: string;
  note_title?: string;
  selection?: string;
  clicked_target?: string;
  cursor_line?: number;
  event_ts_ms?: number;
  metadata?: Record<string, any>;
  received_at?: string;
}

export interface ObsidianContextSessionContract {
  session_id: string;
  user_id: string;
  space_id?: string | null;
  active_note_path?: string | null;
  active_note_title?: string | null;
  active_selection?: string | null;
  recent_events: ObsidianContextEventContract[];
  last_event_at?: string | null;
  plugin_connected?: boolean;
  last_heartbeat_at?: string | null;
  last_heartbeat_age_ms?: number | null;
}

export interface SpaceContract {
  space_id: string;
  name: string;
  owner_id: string;
  status: string;
  monthly_token_budget: number;
  monthly_api_calls: number;
  preferred_model: string;
}

export interface FeatureFlagsContract {
  threading: boolean;
  per_message_model: boolean;
  secure_provider_settings: boolean;
  obsidian_advanced: boolean;
  rpi_compact_layout: boolean;
  obsidian_event_sync: boolean;
}

File: app/contracts.py
"""Versioned API contract models for migration-safe interoperability."""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field

CONTRACT_VERSION = "v1"


class ContextSourceContract(BaseModel):
    type: Literal["FILE", "DIRECTORY", "URL", "API", "DATABASE", "MEMORY", "OBSIDIAN"]
    path: str
    extra: Optional[Dict[str, Any]] = None


class ObsidianContextEventContract(BaseModel):
    event_id: Optional[str] = None
    event_type: Literal["open", "click", "selection", "navigate"]
    note_path: str
    note_title: Optional[str] = None
    selection: Optional[str] = None
    clicked_target: Optional[str] = None
    cursor_line: Optional[int] = None
    event_ts_ms: Optional[int] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)


class ChatRequestContract(BaseModel):
    user_id: Optional[str] = None
    query: str
    model: Optional[str] = None
    provider: Optional[str] = None
    context_sources: Optional[List[ContextSourceContract]] = None
    space_id: Optional[str] = None
    thread_id: Optional[str] = None


class ChatResponseContract(BaseModel):
    user_id: str
    answer: str
    research: Optional[str] = None
    plan: Optional[str] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)


class StatusResponseContract(BaseModel):
    status: str
    required_api_keys: Dict[str, str]
    database_connected: bool
    database_error: Optional[str] = None
    optional_services: Dict[str, bool]
    contract_version: str = CONTRACT_VERSION


class FeatureFlagsContract(BaseModel):
    threading: bool
    per_message_model: bool
    secure_provider_settings: bool
    obsidian_advanced: bool
    rpi_compact_layout: bool
    obsidian_event_sync: bool

